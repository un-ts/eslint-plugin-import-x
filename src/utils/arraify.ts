export const arraify = <T>(value?: T | readonly T[]): T[] | undefined =>
  value ? ((Array.isArray(value) ? value : [value]) as T[]) : undefined
