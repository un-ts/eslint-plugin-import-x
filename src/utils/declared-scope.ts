import type { TSESTree } from '@typescript-eslint/utils'

import type { RuleContext } from '../types'

export function declaredScope(
  context: RuleContext,
  node: TSESTree.Node,
  name: string,
) {
  const references = context.sourceCode.getScope(node).references
  const reference = references.find(x => x.identifier.name === name)
  if (!reference || !reference.resolved) {
    return
  }
  return reference.resolved.scope.type
}
