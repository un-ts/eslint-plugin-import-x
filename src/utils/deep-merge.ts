export type ObjectLike<T = unknown> = Record<string, T>

/**
 * Check if the variable contains an object strictly rejecting arrays
 *
 * @returns `true` if obj is an object
 */
export function isObjectNotArray(obj: unknown): obj is ObjectLike {
  return typeof obj === 'object' && obj != null && !Array.isArray(obj)
}

/**
 * Pure function - doesn't mutate either parameter! Merges two objects together
 * deeply, overwriting the properties in first with the properties in second
 *
 * @param first The first object
 * @param second The second object
 * @returns A new object
 */
export function deepMerge(
  first: ObjectLike = {},
  second: ObjectLike = {},
): Record<string, unknown> {
  // get the unique set of keys across both objects
  const keys = new Set([...Object.keys(first), ...Object.keys(second)])

  return Object.fromEntries(
    [...keys].map(key => {
      const firstHasKey = key in first
      const secondHasKey = key in second
      const firstValue = first[key]
      const secondValue = second[key]

      let value
      if (firstHasKey && secondHasKey) {
        value =
          // object type
          isObjectNotArray(firstValue) && isObjectNotArray(secondValue)
            ? deepMerge(firstValue, secondValue)
            : // value type
              secondValue
      } else if (firstHasKey) {
        value = firstValue
      } else {
        value = secondValue
      }
      return [key, value]
    }),
  )
}
