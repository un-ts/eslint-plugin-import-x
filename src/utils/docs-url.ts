import { version } from '../meta.js'

const repoUrl = 'https://github.com/un-ts/eslint-plugin-import-x'

export const docsUrl = (ruleName: string, commitish = `v${version}`) =>
  `${repoUrl}/blob/${commitish}/docs/rules/${ruleName}.md`
