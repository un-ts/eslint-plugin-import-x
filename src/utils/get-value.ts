import { TSESTree } from '@typescript-eslint/types'

export const getValue = (
  node: TSESTree.Identifier | TSESTree.StringLiteral,
) => {
  switch (node.type) {
    case TSESTree.AST_NODE_TYPES.Identifier: {
      return node.name
    }
    case TSESTree.AST_NODE_TYPES.Literal: {
      return node.value
    }
    default: {
      throw new Error(`Unsupported node type: ${(node as TSESTree.Node).type}`)
    }
  }
}
