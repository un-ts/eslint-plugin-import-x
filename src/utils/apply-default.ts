import { deepMerge, isObjectNotArray } from './deep-merge.js'

/**
 * Pure function - doesn't mutate either parameter! Uses the default options and
 * overrides with the options provided by the user
 *
 * @param defaultOptions The defaults
 * @param userOptions The user opts
 * @returns The options with defaults
 */
export function applyDefault<
  User extends readonly unknown[],
  Default extends User,
>(
  defaultOptions: Readonly<Default>,
  userOptions: Readonly<User> | null,
): Default {
  // clone defaults
  const options = structuredClone(defaultOptions) as AsMutable<Default>

  if (userOptions == null) {
    return options
  }

  // For avoiding the type error
  //   `This expression is not callable. Type 'unknown' has no call signatures.ts(2349)`
  for (const [i, opt] of (options as unknown[]).entries()) {
    if (userOptions[i] !== undefined) {
      const userOpt = userOptions[i]
      options[i] =
        isObjectNotArray(userOpt) && isObjectNotArray(opt)
          ? deepMerge(opt, userOpt)
          : userOpt
    }
  }

  return options
}

type AsMutable<T extends readonly unknown[]> = {
  -readonly [Key in keyof T]: T[Key]
}
