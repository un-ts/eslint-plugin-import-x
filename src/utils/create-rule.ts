import { ESLintUtils } from '@typescript-eslint/utils'

import { docsUrl } from './docs-url.js'

export interface ImportXPluginDocs {
  /** The category the rule falls under */
  category?: string

  recommended?: true
}

export const createRule = ESLintUtils.RuleCreator<ImportXPluginDocs>(docsUrl)
