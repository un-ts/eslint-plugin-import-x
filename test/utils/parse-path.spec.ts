import { parsePath, stringifyPath } from 'eslint-plugin-import-x/utils'

describe('parse-path', () => {
  it('should parse and stringify path expectedly', () => {
    const cases = [
      'foo',
      'foo?query',
      'foo#hash',
      'foo?query#hash',
      'foo#hash?query',
    ]

    for (const input of cases) {
      const output = parsePath(input)
      expect(output).toMatchSnapshot(input)
      expect(stringifyPath(output)).toBe(input)
    }
  })
})
