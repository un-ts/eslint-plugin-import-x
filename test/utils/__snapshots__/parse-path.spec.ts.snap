// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse-path should parse and stringify path expectedly: foo 1`] = `
{
  "hash": "",
  "pathname": "foo",
  "query": "",
}
`;

exports[`parse-path should parse and stringify path expectedly: foo#hash 1`] = `
{
  "hash": "#hash",
  "pathname": "foo",
  "query": "",
}
`;

exports[`parse-path should parse and stringify path expectedly: foo#hash?query 1`] = `
{
  "hash": "#hash?query",
  "pathname": "foo",
  "query": "",
}
`;

exports[`parse-path should parse and stringify path expectedly: foo?query 1`] = `
{
  "hash": "",
  "pathname": "foo",
  "query": "?query",
}
`;

exports[`parse-path should parse and stringify path expectedly: foo?query#hash 1`] = `
{
  "hash": "#hash",
  "pathname": "foo",
  "query": "?query",
}
`;
